doc: update conventions with repl examples and function (in|out)puts

DanielSidhion opened the pull request:
## Description of changes

Closes #276829.

This PR adds two more items to the nixpkgs documentation conventions as discussed in the documentation meetings and PR reviews.

This is related to #278769, where I'm going through the nixpkgs manual to make it more consistent (and also update its content). I'm documenting the conventions I'm applying as I do this work in an effort to:
1. Let other people know the conventions I'm using, should they want to help with this work.
2. Let other people start using the same conventions for future work.
3. Have conventions clearly documented, because they weren't (or didn't even exist) before this work started.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

DanielSidhion committed:
doc: update conventions with repl examples and function (in|out)puts

bzm3r commented:
Regarding function inputs/outputs. I am not so sure about ditching `f :: A -> b -> C` style notation.

Some reasons why it's nice (to me):

* it's very consistent in style, which is helpful when trying to quickly find info
* it's easy to auto-generate (given some doc strings conventions as current RFCs are pushing, or possibly, eventually: a type system)
* it helps communicate that it is very possible to partially evaluate functions. So for instance, if `repeatN :: String -> Integer -> String`, and `"hello" :: String`, we have `repeatN "hello" :: Integer -> String`. Put differently, there are three, equally valid ways to look at `repeatN`: 
1. `repeatN` is a function which takes `String` as input and produces another function `Integer -> String` as output, or 
2. `repeatN` is a function which takes a function `String -> Integer` as input and produces a `String` as output, or 
3. `repeatN` is a function which takes a `String` and `Integer` as input and produces a `String` as output
* it is already a convention used by some Nix IDE tools and documentation
* it is already a convention used by some languages (e.g. Haskell), and similar-ish to the [BNF](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form) convention more generally 

Some reasons why its not nice (to me):

* it takes some getting used to (but its knowledge that is helpful in other languages too)
* nothing is said immediately about what `String` is required (having some way to include variable names along with the types would be helpful...)

But I can imagine a convention where we write along with `runCommand :: String -> AttrSet -> String -> Derivation` something like `runCommand :: name -> drvAttr -> buildCommand -> Derivation` so that we have overall:

```
runCommand :: name -> drvAttr -> command -> Derivation
                        :: String -> AttrSet -> String -> Derivation

Returns a derivation realizing a binary executing specified shell commands. 

name :: String
    The name that Nix will append to the store path in the same way that stdenv.mkDerivation uses its name attribute.

drvAttr :: AttrSet
    Attributes to pass to the underlying call to stdenv.mkDerivation.

buildCommand :: String
    Shell commands to run in the derivation builder.
    
Note
    You have to create a file or directory $out for Nix to be able to run the builder successfully.
```

bzm3r commented on doc/README.md:
```diff
@@ -184,12 +184,64 @@ In that case, please open an issue about the particular documentation convention
 
   Description of what `coolFunction` does.
+
+  ## Inputs
```
Regarding function inputs/outputs. I am not so sure about ditching `f :: A -> b -> C` style notation.

Some reasons why it's nice (to me):

* it's very consistent in style, which is helpful when trying to quickly find info
* it's easy to auto-generate (given some doc strings conventions as current RFCs are pushing, or possibly, eventually: a type system)
* it helps communicate that it is very possible to partially evaluate functions. So for instance, if `repeatN :: String -> Integer -> String`, and `"hello" :: String`, we have `repeatN "hello" :: Integer -> String`. Put differently, there are three, equally valid ways to look at `repeatN`: 
1. `repeatN` is a function which takes `String` as input and produces another function `Integer -> String` as output, or 
2. `repeatN` is a function which takes a function `String -> Integer` as input and produces a `String` as output, or 
3. `repeatN` is a function which takes a `String` and `Integer` as input and produces a `String` as output
* it is already a convention used by some Nix IDE tools and documentation
* it is already a convention used by some languages (e.g. Haskell), and similar-ish to the [BNF](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form) convention more generally 

Some reasons why its not nice (to me):

* it takes some getting used to (but its knowledge that is helpful in other languages too)
* nothing is said immediately about what `String` is required (having some way to include variable names along with the types would be helpful...)

But I can imagine a convention where we write along with `runCommand :: String -> AttrSet -> String -> Derivation` something like `runCommand :: name -> drvAttr -> buildCommand -> Derivation` so that we have overall:

```
runCommand :: name -> drvAttr -> command -> Derivation
                        :: String -> AttrSet -> String -> Derivation

Returns a derivation realizing a binary executing specified shell commands. 

name :: String
    The name that Nix will append to the store path in the same way that stdenv.mkDerivation uses its name attribute.

drvAttr :: AttrSet
    Attributes to pass to the underlying call to stdenv.mkDerivation.

buildCommand :: String
    Shell commands to run in the derivation builder.
    
Note
    You have to create a file or directory $out for Nix to be able to run the builder successfully.
```

DanielSidhion replied:
"ditching" is a strong word - there hasn't been any convention before this, so there's nothing to "ditch".
In any case, I think this comment misunderstands what the convention is about. This is something to provide a common structure to make it easier for people to navigate any part of the manual. It doesn't say how to write the content of each heading.

There is another convention to use definition lists for arguments and attributes, which is also there to provide a common structure for readers. Applied to your example, the `name`, `drvAttr`, and `buildCommand` inputs should ideally be structured in a definition list. That still doesn't force anyone to "ditch" the style notation you like.

DanielSidhion replied:
I updated the PR with another clarification to avoid this kind of misunderstanding again.

bzm3r replied:
I do not think your changes address the question I raised. In the interest of clarity, note that:

* no statements were made regarding an already existing convention. There are a few different "local conventions". A reference was being made to [*one* of these local conventions](https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-runCommand). For clarity, I'll refer to these "local conventions" as "dialects".

* no statements were made regarding personal preference for an existing documentation dialect over another. An attempt was made to express uncertainty around the selection of one dialect over others, without due diligence. 

* (this will be fixed) no statements were made regarding collating a list of existing documentation-dialects, or better yet, to link to existing research on this front.

To fix the last point:
* The `nixos-render-doc` work *might* have had such a collation (not easy to (re)find, therefore, not included for now). 
* [`nixdoc`](https://github.com/nix-community/nixdoc), which is already used for documenting items in [`lib`, generates output like so](https://nixos.org/manual/nixpkgs/unstable/#sec-functions-library-strings)

To fix potential issues around misunderstanding of colloquialisms, note that "to ditch" means:
> To [discard](https://en.wiktionary.org/wiki/discard) or [abandon](https://en.wiktionary.org/wiki/abandon).

I will stick to discard from now on for clarity.

The convention here---even though a work in progress---has the potential to influence new PRs suggesting changing existing `f :: A -> B -> C` structure to use a Input/Output header structure. This is one sense in which `f :: A -> B -> C` style structure may be discarded.

If a convention is ratified by this PR, then it would greenlight changing existing documentation differing from the approved structure. This is another sense in which existing dialects could be discarded.

Therefore, two issues are being raised: 

1. Please do not select a convention, without a summary of existing dialects, and some analysis of their pros and cons. 

2. Please mark clearly that this document is a work in progress that does not endorse any convention yet.

bzm3r commented on doc/README.md:
```diff
@@ -184,12 +184,67 @@ In that case, please open an issue about the particular documentation convention
   # pkgs.coolFunction
 
   Description of what `coolFunction` does.
+
```
I do not think your changes address the question I raised. In the interest of clarity, note that:

* no statements were made regarding an already existing convention. There are a few different "local conventions". A reference was being made to [*one* of these local conventions](https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-runCommand). For clarity, I'll refer to these "local conventions" as "dialects".

* no statements were made regarding personal preference for an existing documentation dialect over another. An attempt was made to express uncertainty around the selection of one dialect over others, without due diligence. 

* (this will be fixed) no statements were made regarding collating a list of existing documentation-dialects, or better yet, to link to existing research on this front.

To fix the last point:
* The `nixos-render-doc` work *might* have had such a collation (not easy to (re)find, therefore, not included for now). 
* [`nixdoc`](https://github.com/nix-community/nixdoc), which is already used for documenting items in [`lib`, generates output like so](https://nixos.org/manual/nixpkgs/unstable/#sec-functions-library-strings)

To fix potential issues around misunderstanding of colloquialisms, note that "to ditch" means:
> To [discard](https://en.wiktionary.org/wiki/discard) or [abandon](https://en.wiktionary.org/wiki/abandon).

I will stick to discard from now on for clarity.

The convention here---even though a work in progress---has the potential to influence new PRs suggesting changing existing `f :: A -> B -> C` structure to use a Input/Output header structure. This is one sense in which `f :: A -> B -> C` style structure may be discarded.

If a convention is ratified by this PR, then it would greenlight changing existing documentation differing from the approved structure. This is another sense in which existing dialects could be discarded.

Therefore, two issues are being raised: 

1. Please do not select a convention, without a summary of existing dialects, and some analysis of their pros and cons. 

2. Please mark clearly that this document is a work in progress that does not endorse any convention yet.

bzm3r commented:
I do not think your changes address the issue I raised. I tried to express this once, but made mistakes. So here's the second try.

In most places, there's a "headerless" style of documentation: e.g. [`nixdoc`](https://github.com/nix-community/nixdoc) generates [output like so](https://nixos.org/manual/nixpkgs/unstable/#sec-functions-library-strings). 

There might be other styles, including headered ones. It's worth understanding which "local styles" are already in use. The `nixos-render-doc` work *might* have had a list of existing styles in use (not easy to (re)find). 

To fix potential issues around misunderstanding of colloquialisms, note that "to ditch" means:
> To [discard](https://en.wiktionary.org/wiki/discard) or [abandon](https://en.wiktionary.org/wiki/abandon).

I will stick to discard from now on for clarity.

The convention as it stands now---even though a work in progress---has the potential to influence new PRs which choose to change existing `f :: A -> B -> C` information to **relying on the Input/Output header structure *alone* to convey information**. (There was already one discussion in the matrix around this.) This is the sense in which `f :: A -> B -> C` information may be discarded.

Therefore:

0. Please provide an explanation of why headers will help readers. Most function documentation styles, in most languages (e.g. Rust or Haskell) do not use headers. They rely on type information alone. to act as headers. 

Some third party Python documentation has subtle headers, but the official ones do not. MDN Javascript documentation does. However, the MDN documentation also has an explicit/separate page *for each* function (not one long scroll-through document, for example like the current Nixpkgs manual). 

1. Please do not select a convention, without a summary of existing function input/output styles, and some analysis of their pros and cons. **(In particular, of *headerless* styles.)**

2. Please consider improving the structure of headerless styles, as an alternative to headers. 

3. Please consider choosing to be clear about what ought to be included under each header. Why leave what goes under each header up for grabs? 

4. Please consider marking clearly that this document is a work in progress that does not endorse any convention yet and does not encourage its use as inspiration just yet.

DanielSidhion commented:
@bzm3r 
In my opinion, the changes I made do address the issue. From what I understood, this seems to be your problem with this PR:
> ...has the potential to influence new PRs which choose to change existing f :: A -> B -> C information to **relying on the Input/Output header structure alone to convey information**

And I have already addressed in the PR that the headers are there for navigational purposes, and they **don't** force the content to be written in any specific way. Anyone can still write `f :: A -> B -> C` in the documentation, but when talking about the inputs specifically, it would be best if they put it under a header. This doesn't mean `f :: A -> B -> C` shouldn't exist. These aren't conflicting things. `f :: A -> B -> C` is content, the headers exist to help navigate the manuals.

The need for headers is way less pronounced in the `nixpkgs.libs` section of the manual (because those are usually short), but it's definitely needed in many other places in the manual, because there are a lot of utility functions that handle a **lot** of stuff behind the scenes.

> Please provide an explanation of why headers will help readers
1. Headers have anchor links that people can use to jump to a specific part of the document, and use in links for other people as well.
2. Manuals are used more as a consultation device. [This section](https://diataxis.fr/reference/#reference-as-description) of the diataxis framework page helps here. Because of that, making things consistent by having headers that are standardised across the manual helps people quickly consult it.

> Most function documentation styles, in most languages (e.g. Rust or Haskell) do not use headers.
The Rust documentation uses headers. I picked a random page there for `std::string::String`. Note how it has "Examples", "Errors", "Panics" (and so on...) headers for a lot of the functions:
- [Examples for new()](https://doc.rust-lang.org/std/string/struct.String.html#examples-1)
- [Examples for with_capacity()](https://doc.rust-lang.org/std/string/struct.String.html#examples-2)
- [Errors for from_utf8()](https://doc.rust-lang.org/std/string/struct.String.html#errors)
- [Examples for from_utf8()](https://doc.rust-lang.org/std/string/struct.String.html#examples-3)

Notably, the Rust documentation doesn't have an "Inputs" header because most of its functions don't have arguments with 10+ different attributes like [dockerTools.buildImage](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/docker/default.nix#L508-L545), for example. Python's functions sometimes have a large number of arguments, and in my opinion it's a really bad experience to navigate their docs. Note that even then, they also have some similarities with the conventions in this PR, for example how they move examples to be the last part of each unit they're documenting: [examples for regular expressions](https://docs.python.org/3/library/re.html#regular-expression-examples), [examples for datetime](https://docs.python.org/3/library/datetime.html#examples-of-usage-datetime).

> 0. Please clarify, in the first post of this PR, what precisely was agreed upon in team meetings as to what the goal of this document is.
Done.

> 3. Please consider improving the structure of headerless styles, as an alternative to headers.
As I wrote already, these are not conflicting things.

> 2. Please do not select a convention, without a summary of existing function input/output styles, and some analysis of their pros and cons. (In particular, of headerless styles.)
I believe I've already addressed that what you're calling a "headerless style" relates to the content of the documentation, not to its structure, and those are not conflicting things.

As for selecting a convention, this is exactly what we're doing in the documentation team meetings. I find an opportunity to standardise something in the nixpkgs manual, then I bring it to the meeting, we discuss it there, and we decide on something. And that is then documented in a PR just like this one.
We shouldn't be spending this much time arguing over the structure of stuff in the nixpkgs manual (and I'm only spending this much time replying to all of this because I think you're being genuine in your concerns and I respect that).

The good thing about the structure of something is that it can easily be changed later (as opposed to the content, which is way harder). If we find out later that some other structure is better, we can quickly move to it. We might need to adapt some text in the process, but we won't need to completely change the content of the documentation.

(and I already know this is inefficient and so on, all of this is already addressed in #278769 - I'm the one doing the work to bring consistency to the manuals, so I'll work this way, even though it seems inefficient, because it's what helps me make progress on the content as quickly as possible, which is ultimately what matters)

> 4. Please consider choosing to be clear about what ought to be included under each header. Otherwise, please explain why this document would prefer to leave what goes under each header up for grabs?
It should be up to whoever's writing the documentation to figure out what content goes where, and even whether they should stick to the convention or not. I can't predict all the possible content that goes in the manuals, but I do know that a good part of it documents utility functions, and those follow roughly the same documentation structure. Hence why a convention for following the same structure (if it makes sense, which is also to be determined by whoever's writing the documentation).

I thought this was kind of implied in the conventions, but I'll update the PR to make it clear that the conventions are about keeping a consistent structure in the manuals.

> 5. Please consider marking clearly that this document is a work in progress that does not endorse any convention yet and does not encourage its use as inspiration just yet.
This document does encourage people to use the conventions laid out in it. That's (also) why they're being documented.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-01-15-documentation-team-meeting-notes-104/38305/1

bzm3r commented:
@DanielSidhion Okay: I will step out of your way then, and not raise further issues here on this point. 

If in the future I feel that that the Input/Output header style does not work well, I'll create a separate PR to address that.

> We shouldn't be spending this much time arguing over the structure of stuff in the nixpkgs manual (and I'm only spending this much time replying to all of this because I think you're being genuine in your concerns and I respect that).

Could you update the first post of this PR again with what was discussed in [this documentation team meeting](https://discourse.nixos.org/t/2024-01-15-documentation-team-meeting-notes-104/38305/1)?

For now, some final thoughts that are mostly just notes for myself:
* structure of content is essential, while cosmetics of their presentation is secondary (and malleable)
* linkability and searchability of documentation are key in making it "map-like"
* headers are not necessary for content to be linkable
* content *can* be a header (for example: in Rust, the function signature)
* if a header does not promise content of certain value, and consistently deliver on that promise, it will become useless to readers

infinisil commented on doc/README.md:
```diff
@@ -184,12 +187,67 @@ In that case, please open an issue about the particular documentation convention
+- When documenting functions or anything that has inputs/outputs and example usage, use nested headings to clearly separate inputs, outputs, and examples.
+  Keep examples as the last nested heading, and link to the examples wherever applicable in the documentation.
+
+  The content of the nested headings is still up to the writer.
```
I feel like this is so obvious that we don't need to mention it

Though hopefully in the future the input attribute docs would be automatically generated by nixdoc ;)

infinisil commented:
Reviewed in the docs team meeting, looking good!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-01-25-documentation-team-meeting-notes-106/38792/1
